### PR TITLE
make ReinterpretArray more Offset-safe

### LINKED ISSUE #44243: Vector indexing of wrappers of ReshapedReinterpretArray fails
I've tried it with `OffsetArrays`, but I guess the behavior would persist with other wrappers. I'm not totally sure what causes this, as I've not looked into it.

```julia
julia> A = reinterpret(reshape, Float64, ones(ComplexF64, 3, 3))
2×3×3 reinterpret(reshape, Float64, ::Matrix{ComplexF64}) with eltype Float64:
[:, :, 1] =
 1.0  1.0  1.0
 0.0  0.0  0.0

[:, :, 2] =
 1.0  1.0  1.0
 0.0  0.0  0.0

[:, :, 3] =
 1.0  1.0  1.0
 0.0  0.0  0.0

julia> B = OffsetArray(A)
2×3×3 OffsetArray(reinterpret(reshape, Float64, ::Matrix{ComplexF64}), 1:2, 1:3, 1:3) with eltype Float64 with indices 1:2×1:3×1:3:
[:, :, 1] =
 1.0  1.0  1.0
 0.0  0.0  0.0

[:, :, 2] =
 1.0  1.0  1.0
 0.0  0.0  0.0

[:, :, 3] =
 1.0  1.0  1.0
 0.0  0.0  0.0

julia> B[:, 1, 1]
ERROR: MethodError: no method matching _maybe_reshape(::Base.IndexSCartesian2{2}, ::OffsetArray{Float64, 3, Base.ReinterpretArray{Float64, 3, ComplexF64, Matrix{ComplexF64}, true}}, ::Base.Slice{OffsetArrays.IdOffsetRange{Int64, Base.OneTo{Int64}}}, ::Int64, ::Int64)
Closest candidates are:
  _maybe_reshape(::Base.IndexSCartesian2, ::Base.ReinterpretArray{T, N, S, A, true} where {T, N, S, A<:(AbstractArray{S})}, ::Any...) at ~/Downloads/julia/julia-1.7/share/julia/base/reinterpretarray.jl:255
  _maybe_reshape(::IndexLinear, ::AbstractArray, ::Any...) at ~/Downloads/julia/julia-1.7/share/julia/base/multidimensional.jl:842
  _maybe_reshape(::IndexCartesian, ::AbstractArray, ::Any...) at ~/Downloads/julia/julia-1.7/share/julia/base/multidimensional.jl:844
  ...
Stacktrace:
 [1] _getindex
   @ ./multidimensional.jl:839 [inlined]
 [2] getindex(::OffsetArray{Float64, 3, Base.ReinterpretArray{Float64, 3, ComplexF64, Matrix{ComplexF64}, true}}, ::Function, ::Int64, ::Int64)
   @ Base ./abstractarray.jl:1218
 [3] top-level scope
   @ REPL[17]:1

julia> A[:, 1, 1]
2-element Vector{Float64}:
 1.0
 0.0
```

This works on the other hand:

```julia
julia> C = @view A[:, :, :]
2×3×3 view(reinterpret(reshape, Float64, ::Matrix{ComplexF64}), :, :, :) with eltype Float64:
[:, :, 1] =
 1.0  1.0  1.0
 0.0  0.0  0.0

[:, :, 2] =
 1.0  1.0  1.0
 0.0  0.0  0.0

[:, :, 3] =
 1.0  1.0  1.0
 0.0  0.0  0.0

julia> C[:, 1, 1]
2-element Vector{Float64}:
 1.0
 0.0

julia> IndexStyle(C)
IndexCartesian()
```

### PR DESCRIPTION
**Pull request by adienes**
fixes https://github.com/JuliaLang/julia/issues/44243 (and adds more tests besides)

### CODE REVIEW

File: test/reinterpretarray.jl [hunk 1]
```diff
@@ -37,10 +39,6 @@ test_many_wrappers(B, (identity, tslow)) do _B
     @test @inferred(size(reinterpret(reshape, Int128, _B))) == (3,)
 end
 
-test_many_wrappers(C) do Cr
```
Comment (adienes) [inline test/reinterpretarray.jl:40]: just folding this into the same at L163

**Comment by adienes** [comment]
I tried also adding a wrapper test for an offset array with non-zero offsets. I think the implementation is actually correct (besides what's fixed in this PR) and the tests would pass, but a lot of the tests check equality against a reference array which are no longer valid if the axes differ, so it requires a larger / more annoying refactoring to the test suite here.

**Comment by adienes** [comment]
one thing to note is that [L91](https://github.com/JuliaLang/julia/pull/58898/files#diff-abc7e7319cd622343d5befe349fd3af23960c7b66424e7a0edb14e6a33c9bab7L91) actually makes the defensive `first(axes(A, 2)) + ind.j - 1` here unnecessary and we could probably do a bare `ind.j`. but in the hypothetical case where in the future `ReinterpretArray` is made more offset-friendly, it would be easy to forget to update that.

another option would be to make the implementation `ind.j` for now but add a test to catch the hypothetical above

given that this expands the test coverage to two full new wrappers, only changes erroring paths into non-errors (and does not touch any non-erroring paths), and the implementation seems pretty uncontroversial to me, absent protest I'll merge this by the end of the week
